### PR TITLE
Add startup/shutdown dependencies and dependency caching lifespan control

### PR DESCRIPTION
**Pull request by adriangb**
## Issues this addresses
- #2057
- #617
- #2697 

To address these issues, this PR provides for more advanced dependency injection concepts.
In particular:
- Gives users the ability to set a cache scope so that values can be cached for the app's lifetime instead of just a single request.
- Give users control over when teardown is run for dependencies with yield. It can now be run during app shutdown, after the response is sent (current default) or right before the response is sent (#2697)

Finally, this PR gives lifespan events (startup/shutdown) the dependency injection.

These changes are presented together for cohesiveness (they aren't very useful individually), but they can easily be split into multiple smaller PRs for easier review.

Here's a small motivating example:

```python
from asyncpg import create_pool, Connection, Pool
from fastapi import Depends, FastAPI
from pydantic import BaseSettings


class Config(BaseSettings):
    dsn: str


def get_config() -> Config:
    return Config()


async def get_connection_pool(cfg: Config = Depends(get_config)) -> Pool:
    async with create_pool(dsn=cfg.dsn) as pool:
        yield pool


async def get_connection(pool: Pool = Depends(get_connection_pool, lifetime="app", cache="app")) -> Connection:
    async with pool.acquire() as conn:
        yield conn


async def check_db_connection(conn: Connection = get_connection()) -> None:
    assert (await conn.execute("SELECT 1;")) == 1


app = FastAPI(on_startup=[check_db_connection])


@app.get("/")
async def root(conn: Connection = get_connection()) -> int:
    return await conn.execute("SELECT 2;")  # or some other real query

```

### CODE REVIEW

File: fastapi/applications.py
Comment (adriangb) [inline fastapi/applications.py:99]: This stuff I'm a bit unclear about. I'm pretty sure the desired behavior is indeed for these dependencies to be re-initialized on every app startup, but I'm not sure if there's any precedent in the codebase for something like this. This is what I came up with. Open to ideas.

File: fastapi/applications.py
Comment (adriangb) [inline fastapi/applications.py:101]: Clearing the dependencies at shutdown aught to help with garbage collection by releasing the refs.

**Comment by adriangb** [comment]
@Faylixe @sm-Fifteen, based on your input in the linked issues, I'd appreciate if you could glance over this and give any feedback. Thanks!

**Comment by adriangb** [comment]
> This even supports sync and async yield dependencies, so that you can easily do startup/shutdown handling

Realized this isn't working. Will think of how to fix it, I hope it's doable

File: fastapi/dependencies/models.py
Comment (Mause) [inline fastapi/dependencies/models.py:36]: Why not add a `Literal["request", "app"]` here?

File: fastapi/applications.py
Comment (Mause) [inline fastapi/applications.py:135]: Rather than repeating the same type annotation over and over, define an alias?

File: fastapi/applications.py
Comment (Mause) [inline fastapi/applications.py:101]:
```suggestion
        self.on_event("shutdown")(self.lifetime_dependencies.clear)
```

File: fastapi/applications.py
Comment (Mause) [inline fastapi/applications.py:141]: What's the reason behind this getter and setter? Why not just assign the variable in the `__init__` method?

**Review by Mause — CHANGES_REQUESTED** [review]

File: docs/en/docs/tutorial/dependencies/lifetime-dependencies.md
Comment (adriangb) [reply docs/en/docs/tutorial/dependencies/lifetime-dependencies.md:6]: I'm referring to a class name. Shouldn't it be camel cased?

File: fastapi/applications.py
Comment (adriangb) [reply fastapi/applications.py:135]: That sounds like a good idea.

File: fastapi/dependencies/models.py
Comment (adriangb) [reply fastapi/dependencies/models.py:36]: I made an alias `DependencyLifetime = Literal["request", "app"]`

File: fastapi/applications.py
Comment (adriangb) [reply fastapi/applications.py:141]: Probably works as well, what I have here is over complicated and doesn't fully work.

Really I'm just struggling to come up with a good way to deal with binding the lifetime dependency stack `__aenter__`/`__aexit__` to the app lifetime. Do you have any suggestions in that regard?

File: docs/en/docs/tutorial/dependencies/lifetime-dependencies.md
Comment (Mause) [reply docs/en/docs/tutorial/dependencies/lifetime-dependencies.md:6]: `Database` is one word, not two

Comment (adriangb) [reply docs/en/docs/tutorial/dependencies/lifetime-dependencies.md:6]: Fair enough, I'll apply the suggestion.

Comment (adriangb) [reply docs/en/docs/tutorial/dependencies/lifetime-dependencies.md]: Pushed the change. Thank you for the suggestion!

File: fastapi/applications.py
Comment (adriangb) [reply fastapi/applications.py]: Here is the alias I made, also the one referred to in https://github.com/tiangolo/fastapi/pull/3516#discussion_r667614049: 
https://github.com/tiangolo/fastapi/blob/cecd77fdd12ac58c66631e9abac25e8c5149eccc/fastapi/dependencies/models.py#L7-L11

File: fastapi/applications.py
Comment (adriangb) [reply fastapi/applications.py]: I completely reworked this part. I got rid of the setters/getters, and it is now handling lifespans as expected. However, I'm not convinced that creating 3 async closures in `__init__` and binding them to `on_start`/`on_shutdown` is the cleanest way to do this.

File: fastapi/params.py
Comment (adriangb) [inline fastapi/params.py:5]: I was getting circular imports, I'm not via what path. Since these are just type hints, this solves it, but I'm open to moving the aliases or another solution.

**Comment by sm-Fifteen** [comment]
I'm not personally sold on having the dependency's lifespan be specified on the caller's end instead of the dependency's end. On one hand, it means any dependency marked as a lifetime dep anywhere causes all of its transitive dependencies to be run and cached as lifetime deps at the same time, which may not be what the user intended. It's also unclear to me what happens if the same function ends up called both as a request-scoped and a lifetime-scoped dependency at different places throughout the application. Would the request-scoped dep short-circuit to using the already available lifetime dep or would it create a separate instance of it? What if the lifetime dep hasn't been initialized yet?

> The main difference between regular startup events and this system is that this system is lazy: nothing is run until the first request that needs it comes in.
> 
> Or more generally, this is not DI for startup/shutdown events (which is what the above linked issues were asking for) but rather startup/shutdown events for the DI system. Maybe it's not what we wanted, but perhaps it's just as good?

More generally, I'm not sure of the benefit of lazily-loaded shared resources compared to having dependencies context-managed by the lifetime protocol (the main reason I can think of being parametrized dependencies, which weren't the sort of thing I was considering using for lifetime deps in the first place).

> In most real world scenarios, the first request will come from a load balancer, Kubernetes, etc to a healthcheck endpoint, and the app won't even take outside traffic until it consistently is returning 200.
If one wanted to, they could make a Startup dependency that collects all lifetime dependencies and binds them to the healthcheck to ensure that they get initialized before any user requests:

I can't speak for most FastAPI users, but the sort of setup I'm using (just a plain reverse proxy for a small API server connected to a bunch of heterogeneous databases) wouldn't fit under that scenario.

Really, my main use case for lifetime deps when I opened #617 was to not have active object handles littering the global namespace outside of functions (see https://github.com/tiangolo/fastapi/issues/726#issuecomment-557687526, and imagine something like that, but minus the underscored global variables). I figured those could just be wrapped in CM-like functions and annotated for FastAPI to know that those functions, when used as dependencies, are not supposed to be run, but to have the eagerly singleton'd result (hopefully a *thread-safe* result, I don't think that can reliably be checked by FastAPI, so it's up to the user to make sure) returned instead.

It also generally makes sense to me that your server should fail to startup if the database connection your application requires to run is unreachable, or if your core config file can't be found, which is why I would have imagined these as not being lazy, but we might just have different use cases for these.

**Comment by adriangb** [comment]
> we might just have different use cases for these

This is very possible, I may have been shortsighted with my use cases, but I'll try to answer some of good questions you bring up

> It's also unclear to me what happens if the same function ends up called both as a request-scoped and a lifetime-scoped dependency at different places throughout the application

> Would the request-scoped dep short-circuit to using the already available lifetime dep or would it create a separate instance of it?

I think this is an important question. To me it boils down to "is an app lifespan dep == to the same dep but with a request lifespan?". Let me know if this interpretation is incorrect.
We can make it work either way, I think we'd just have to add a tuple element to the cache key to make them unequal.
In other words, we can have either of these two work:

```python
class Config:
    ...


def get_config() -> Config:
    return Config()


@app.get("/")
def root(cfg1 = Depends(get_config, lifetime="app"), cfg2 = Depends(get_config)):
    assert cfg1 is cfg2  # option 1, current implementation
    assert cfg1 is not cfg2  # option 2, adding the `lifetime` parameter as a cache key element
```

> On one hand, it means any dependency marked as a lifetime dep anywhere causes all of its transitive dependencies to be run and cached as lifetime deps at the same time, which may not be what the user intended

The intention is that only the top level (let's call it root) dependency gets marked as a lifetime dependency and cached as such. If a transitive dependency (`get_config` in this example) is used by both a lifetime dependency and a non-lifetime dependency it will only be "shared" the first call:

```python
class Config:
    ...


def get_config() -> Config:
    return Config()


def lifetime(cfg: Config = Depends(get_config)):
    return cfg

def request(cfg: Config = Depends(get_config)):
   return cfg

calls = 0

@app.get("/")
def root(cfg1 = Depends(lifetime, lifetime="app"), cfg2 = Depends(request)):
    global calls
    calls +=1
    if calls == 1:
        assert cfg1 is cfg2
    else:
        assert cfg1 is not cfg2
```

>  I'm not sure of the benefit of lazily-loaded shared resources compared to having dependencies context-managed by the lifetime protocol 

I agree, it would be nice to have the opposite (DI for the startup/shutdown system) as well. Like you say, it often makes sense to not even start the application unless the database, etc. are able to connect. But maybe there is a use case for both?

> just a plain reverse proxy for a small API server

As far as I know, even simple reverse proxies like NGINX have healthcheck functionality.

In your use case, how do you know a deploy was successful if you don't have some sort of healthcheck / rediness?

> I'm not personally sold on having the dependency's lifespan be specified on the caller's end instead of the dependency's end

This is not necessarily a solution to this, but maybe something like the following can be used to clean up things:

```python
def DBConnection() -> Connection:
    def real_dep(cfg: Config = Depends(get_config)):
        ...
    return Depends(real_dep, lifetime="app")


@app.get("/")
def root(conn: Connection = DBConnection()):
    ...
````

Now the caller doesn't have to specify `Depends(..., lifetime="app")`.

**Comment by Faylixe** [comment]
> I'm not personally sold on having the dependency's lifespan be specified on the caller's end instead of the dependency's end. 

+1 on this, not a big fan as well.

> Now the caller doesn't have to specify Depends(..., lifetime="app").

In that case, if dependency behavior should be defined by the user I would prefer having a factory function handling this:

```python
def AppDepends(*args, **kwargs) -> Any:
    return Depends(*args, **kwargs, lifespan="app")
```

Which then can be use like a traditional one:

```python
@app.get("/")
def root(cfg1 = AppDepends(get_config), cfg2 = Depends(get_config)):
```

**Comment by adriangb** [comment]
Yeah, I totally understand on:
- Wanting this to be tied to startup/shutdown
- Wanting to specify the lifetime on the dependency side

The _main_ reason I'm not doing it like that is that:
- There is no DI integration into startup/shutdown
- The DI system currently relies on specifying things on the caller's side (eg. `use_cache`)

And so doing it this way allowed me to do it in relatively few lines of code, with no major refactors, while still satisfying all of _my_ needs.
I think trying to hit either of those other two features would require more extensive refactors (although maybe I'm just not familiar enough with the codebase).

I do think there's some use for both, I can see scenarios where it is desirable to lazily initialize lifetime dependencies.

**Comment by graingert** [comment]
hello! I'd highly recommend using the `Starlette(lifespan=...)` interface for this so people can create `anyio.TaskGroup`s in their lifespan dependencies.

To support this, `lifespan` was changed in 0.16.0 to only support async context manager factories, eg using `@contextlib(2).asynccontextmanager`, and importantly the `yield` statement can now throw an instance of `anyio.get_cancelled_exc_class()` eg:


```python
@contextlib.asynccontextmanager
async def TaskGroup() -> anyio.TaskGroup:
    async with anyio.create_task_group() as tg:
        yield tg  <- this might throw asyncio.CancelledError or trio.CancelledError!
```

**Comment by adriangb** [comment]
100% agreed, thank you for the suggestion.

The only reason I didn't use `Starlette(lifespan=...)` is because FastAPI does not support that yet (in particular, FastAPI's router class doesn't). And so adding those keyword arguments to `Router`, etc. would have been out of scope for this PR. It may be possible to skirt around adding any public parameters, but there's an even bigger problem: because Statlette can't have both `Starlette(startup=[...])` and `Starlette(lifespan=...)` at the same time, if we add a `lifespan` argument we would disable users existing startup/shutdown hooks, which would break a lot of code.

**Comment by sm-Fifteen** [comment]
> The second bigger problem is that Statlette can't have both `Starlette(startup=[...])` and `Starlette(lifespan=...)` at the same time, if we add a `lifespan` argument we would disable users existing startup/shutdown hooks, which would break a lot of code.

The startup and shutdown hooks are merely shimmed by the default starlette lifespan context manager (see encode/starlette#799), now. There's nothing stopping FastAPI from adding a custom lifespan handler to take care of lifespan dependencies that also shims the old startup and shutdown events.

**Comment by adriangb** [comment]
@graingert I pushed a version using `Starlette(lifespan=...)` in 
629e465
@sm-Fifteen you're right, thanks for the recommendation. I ended up with this shim: https://github.com/tiangolo/fastapi/blob/629e465d7afe8fac442066c62f2897455f5198f8/fastapi/routing.py#L454-L479 I think the only tricky bit left is that whenever FastAPI enables `FastAPI(lifespan=...)` we'll have to handle adding the user's lifespan into this glue lifespan (I suppose checking if it is a context manager or generator, etc.).

File: docs_src/dependencies/tutorial013.py
Comment (graingert) [inline docs_src/dependencies/tutorial013.py:26]:
```suggestion
@contextlib.asynccontextmanager
def DBConnection(config: AppConfig = Depends(Config)) -> Generator[Database, None, None]:
```

Pushing the responsibility of managing the generator down to the user solves a number of weird edge cases around Exceptions/StopIteration and Traceback preservation

Comment (adriangb) [reply docs_src/dependencies/tutorial013.py]: makes sense, thanks!

**Comment by graingert** [comment]
I think it's worth waiting for the Starlette 0.16.0 upgrade in this pr because so much changed about how lifespan works

**Comment by adriangb** [comment]
Are there more changes beyond having to wrap the lifespan func in `@contextlib.asynccontextmanager`?

Do you know if FastAPI has a plan / upgrade path for Starlette versions?

**Comment by adriangb** [comment]
Looks like https://github.com/tiangolo/fastapi/pull/3372 is what we'd be waiting for

**Comment by adriangb** [comment]
Alrighty folks, some exciting news, I reworked this to support dependencies in startup/shutdown events (and also in lifespan managers, whenever we update to `starlette=0.16.0`)

I decided to keep this in the same PR since:
1. The underlaying implementation using the lifespan AsyncExitStack is shared.
2. The question of caching/sharing deps between lifespan/startup events and endpoints naturally arises, and is addressed using the original mechanism of this PR.

To elaborate on (2), I renamed the parameter `Depends(..., lifespan="app")` to `Depends(..., cache_lifespan="app")`, which not only more accurately reflects what it is doing, but also gives users the ability to choose if they want to cache startup/lifespan dependencies to be shared with endpoints or not. Endpoints also always have the option of using `Depends(..., cache=False)` to independently disable caching of a particular dependency for a particular endpoint.

I still need to update the docs, but I have basic tests for most of the behavior described above.

**Comment by sm-Fifteen** [comment]
The wording change goes a long way to make the changes fit more cleanly in the current model, since they would allow deprecating/replacing the current `use_cache` parameter and establishes a clear mental model of what should happen if the same dependency is called with or without that parameter (or with different values). While it doesn't change my reservations about how caller-side vs function-side lifespan specifications, this has the potential to work well enough in more varied use cases and would be easy to tutorialize on top of the existing dependency stuff, so I'm willing to give it a go.

A few things I should note, though:
* Like with all forms of shared resources (whether passed as dependencies or from global scope), users should ensure that any resource managed this way is both thread-safe and coroutine-safe (one does not always imply the other, see the issues raised with `scoped_session` in https://github.com/tiangolo/fastapi/issues/726#issuecomment-868750382)
* Like I mention in the "Additional context" part of my original post for https://github.com/tiangolo/fastapi/issues/617#issue-506258325, the user should be aware that the shutdown phase is not strictly guaranteed to run if the server doesn't get to exit cleanly (so that they can make sure that database connections time out properly and such).
* Starlette currently has a problem with lifetime on mounted subapplications (encode/starlette#649) so you might want to either have a backup mechanism in case lifetime didn't fire properly or somehow disallow lifetime deps on sub-applications somehow.

**Comment by graingert** [comment]
Thanks for your work on this feature, I'm really exited to use it.

There's some issues with lifespan and multiple threads exacerbated by the starlette.TestClient: https://github.com/django/asgiref/pull/277

And there's some additional issues with Trio and AsyncExitStack: https://github.com/python-trio/trio/issues/1243

**Comment by graingert** [comment]
> the user should be aware that the shutdown phase is not strictly guaranteed to run if the server doesn't get to exit cleanly

This should no longer be true with Starlette 0.16.0. if the lifespan startup hook completes the lifespan shutdown hook will certainly be called. (Unless the thread is aborted with os._exit or get_event_loop().run_until_complete is used)

**Comment by adriangb** [comment]
> The wording change goes a long way to make the changes fit more cleanly in the current model

👍 glad it helps!

> allow deprecating/replacing the current use_cache

Hmm I was thinking these would work in tandem. One establishing the lifespan, another if caching should be used at all or not.

> users should ensure that any resource managed this way is both thread-safe and coroutine-safe

Agreed. I think there's not much we can do about this other than write good documentation, right?

> user should be aware that the shutdown phase is not strictly guaranteed to run if the server doesn't get to exit cleanly

Agreed again. But also documentation right?

If I am misunderstanding and either of these have implementation implications, please let me know.

> Starlette currently has a problem with lifetime on mounted subapplications (...) you might want to either have a backup mechanism in case lifetime didn't fire properly

This seems like something where we should let Starlette resolve it (maybe holding off on implementing this feature until they do if it's a deal breaker), not something FastAPI should be working around.

**Comment by graingert** [comment]
>  If those two parts could be welded wold put back together, I feel like we'd have a much more solid mechanism for managing lifetime-scoped ressources than the current one. 

Afaik this is exactly the fix I did here https://github.com/encode/starlette/pull/1227

**Comment by adriangb** [comment]
> There's some issues with lifespan and multiple threads exacerbated by the starlette.TestClient

> And there's some additional issues with Trio and AsyncExitStack

To be honest @graingert, these are a bit too deep into ASGI and anyio for me to fully understand. If you have the time, I would appreciate a bit of guidance on the implication to this PR.

> > If those two parts could be welded wold put back together, I feel like we'd have a much more solid mechanism for managing lifetime-scoped resource than the current one.

(from https://github.com/tiangolo/fastapi/pull/3516#issuecomment-884333414)

I couldn't find where this quote is coming from, maybe it was from an edited comment? Could you please link me to the context? Thanks

**Comment by graingert** [comment]
https://github.com/tiangolo/fastapi/issues/617#issuecomment-548465697

**Comment by adriangb** [comment]
Thank you @graingert

I do agree that the `give me a context manager as a lifespan handler` approach is much, much better than startup/shutdown events. And, as far as I can tell, this PR will be compatible with it with minimal changes (once FastAPI updates it's Starlette version).

**Comment by sm-Fifteen** [comment]
> https://github.com/tiangolo/fastapi/issues/617#issuecomment-548465697
> > If those two parts could be welded wold put back together, I feel like we'd have a much more solid mechanism for managing lifetime-scoped ressources than the current one.
> 
> Afaik this is exactly the fix I did here [encode/starlette#1227](https://github.com/encode/starlette/pull/1227)

Ah, I thought encode/starlette#799 had already managed this, but I guess forcing the use of context managers instead of generators could make everything more robust in a bunch of cases (hard to say from the code diff alone).



> > the user should be aware that the shutdown phase is not strictly guaranteed to run if the server doesn't get to exit cleanly
> 
> This should no longer be true with Starlette 0.16.0. if the lifespan startup hook completes the lifespan shutdown hook will certainly be called. (Unless the thread is aborted with os._exit or get_event_loop().run_until_complete "Pending task was destroyed!" is used)

That's what I was referring to. VSCode will immediately kill your process [if stopped via the debugger's stop button](golang/vscode-go#120) (I'm not sure whether or not the Python extension still does it), and I've had that issue bite me more than once. Non-graceful shutdowns happen, so I figure that's something to be aware of.


> > allow deprecating/replacing the current use_cache
> 
> Hmm I was thinking these would work in tandem. One establishing the lifespan, another if caching should be used at all or not.

Ah, I figured that's what ~~`cache_lifespan=None`~~ `cache_lifespan=False` (`None` seems like a better parameter type to use) was supposed to be for.

> > users should ensure that any resource managed this way is both thread-safe and coroutine-safe
> 
> Agreed. I think there's not much we can do about this other than write good documentation, right?
> 
> > user should be aware that the shutdown phase is not strictly guaranteed to run if the server doesn't get to exit cleanly
> 
> Agreed again. But also documentation right?

Yes, both of these were highlighted more in regards to documentation than to code, given how there's not really anything that can realistically be done at runtime for either of those issues.

> > Starlette currently has a problem with lifetime on mounted subapplications (...) you might want to either have a backup mechanism in case lifetime didn't fire properly
> 
> This seems like something where we should let Starlette resolve it (maybe holding off on implementing this feature until they do if it's a deal breaker), not something FastAPI should be working around.

Yeah, I'm not sure what to do in that regard here either. On one hand, it may lead to unexpected application breakage, on the other hand I don't know if that use-case is enough of a roadblock to put this feature on hold until the upstream bug gets resolved, given it's been sitting untouched for almost 2 years.

**Comment by adriangb** [comment]
> I figured that's what cache_lifespan=None cache_lifespan=False (None seems like a better parameter type to use) was supposed to be for.

I actually like the idea of replacing `cache=False` with `cache_lifespan=None`. But I think it would break backwards compatibility / we'd have to figure out what to do if `cache_lifespan=None, cache=True`.

**Comment by graingert** [comment]
Seems like something akin to pytest.fixture scope might work? Maybe with an Enum instead?

```
class Scope/(enum.Enum):
    lifespan = enum.auto()
    request = enum.auto()
```

?

**Comment by sm-Fifteen** [comment]
> I actually like the idea of replacing `cache=False` with `cache_lifespan=None`. But I think it would break backwards compatibility / we'd have to figure out what to do if `cache_lifespan=None, cache=True`.

From what I've seen elsewhere, you usually check for the new parameter (`cache_lifespan=None`) first and use it if it's set, and then either ignore the old one, emit a deprecation warning or throw if both happen to be set. Given how `use_cache` currently has a default value of `True`, `cache_lifespan` could default to `Ellipsis` (unset, aliasing to `'request'` by default) so FastAPI can check if it has been set intentionally or not, and have it take priority over `use_cache` if it's is set (by the application programmer) to something else.

**Comment by adriangb** [comment]
> Seems like something akin to pytest.fixture scope might work? Maybe with an Enum instead?

That could work, but it would add an extra import.

In any case, it seems like the use of enum and/or replacing `cache: bool = {True,False}` seem totally doable, albeit minor things that don't change the fundamentals of the proposed change. I'm going to hold off on those changes until we can get review from @tiangolo  / more clarity on the viability of the PR in general.

**Comment by graingert** [comment]
This documentation goes into more details about the sort of shared startup and shutdown that I think FastAPI will need to implement: https://pytest-trio.readthedocs.io/en/stable/reference.html#an-important-note-about-yield-fixtures

**Comment by adriangb** [comment]
> This strategy of using a task for each fixture (read Dependency lifespan) has a number of benefits, it means task group and cancel scopes work as expected and it means that you can execute startup and shutdown handlers concurrently

Interesting. I will have to do a bit more reading / give this some more thought, I'm not super familiar with some of this stuff, especially anyio (I'm just a vanilla asyncio user). If you prefer, please feel free to jump in and commit changes instead of holding my hand.

**Comment by graingert** [comment]
It looks like something that can be handled by anyio for you. Eg you pass it a DAG of async context manager factory functions and you get back a single async connect manager that schedules them all for you

**Comment by adriangb** [comment]
On another note, I pushed doc/example updates (just examples, I still have to rewrite the docs). There are now 2 examples, that are really just rewrites of current examples:
- A rewrite of the [Settings as a dependency](https://fastapi.tiangolo.com/advanced/settings/?h=sett#settings-in-a-dependency) example that replaces `@lru_cache` with a `cache_lifespan ="app"` dependency.
- A rewrite of the [Async Databases example](https://fastapi.tiangolo.com/advanced/async-sql-databases/?h=async+data) that instead of hardcoding the DB URL as a global variable pulls it from a Pydantic config (this is typical for a real world use I think) and connects to the database as a startup event. The startup event dependency is marked with `cache_lifespan="app"` so that endpoints get the same instance of the result, and context management of the database connection is handled via the app's lifespan AsyncExitStack. It then uses dependency overrides to override the Postgres URL with an SQLite url (the original example just comments out the global variable and replaces it)

**Comment by GabeMedrash** [comment]
Watching from afar with keen interest...

May also be helpful to see how other web frameworks have approached pieces of this question. Spring, for example, is an interesting comparison. Related to dependency scoping, it allows for beans--depending on the application's context--to have one of 6 scopes: https://www.baeldung.com/spring-bean-scopes. Perhaps "request" and "app" scoping is enough for FastAPI for now, but, may be helpful to see how this feature has grown in a longer-standing framework.

**Comment by adriangb** [comment]
That's very interesting, thank you for sharing.

It seems to me that, as currently implemented, this PR provides the equivalent of several of those scopes:
- `cache=True, cache_lifetime="request"`: equivalent to Spring's `Request` and `WebSocket` scopes.
- `cache=False, cache_lifetime="request"`: equivalent to Spring's `Prototype` scope.
- `cache=True, cache_lifetime="app"`: equivalent to Spring's `Singleton` scope.
- `cache=False, cache_lifetime="app"`: this one doesn't make much sense. If used in an endpoint, it would just keep adding to the exit stack every time it's called, and they would all exit when the app stops. I'd lean towards explicitly prohibiting this behavior.

There is no real equivalent for Springs `Application` scope, if I'm understanding it correctly. You'd be better off just using a module level global.

@GabeMedrash please correct me if this interpretation is flawed (it probably is in some way, I'm not a Spring user).

So maybe we should just have one parameter like `cache={request/True,prototype/False,app}`?

**Comment by GabeMedrash** [comment]
Yeah, that's a good point: there's a decent amount of flexibility built into the combination of `cache` and `cache_lifespan` as you've implemented here, and your comparisons look about right to me on first glance, though I'm sure there are "details" (there always are with Spring). 

Of course, the challenge is making sure that it's clear/explicit how to accomplish a task given the API. I'm just some random guy on the internet--no need to give me too much attention--but, I tend to find, e.g., `scope="prototype"` a little clearer than `cache=False, cache_lifespan="request"`. Perhaps though something like `scope="prototype"` (and the like) is ultimately syntactic sugar over these more low-level args.

**Comment by adriangb** [comment]
I think you're making a very good point in that `scope` (in the Spring sense) is syntactic sugar on top of a combination of how long the value is stored (`cache` in our case) and when the bean is created/destroyed (`lifespan/lifetime` in our case).

I think the nice thing is that because of how generator dependencies and the cache are implemented here, we have full control over both aspects independently. This got me thinking about all of the options we have that users might want. I'm going to leave aside _creation_ of dependencies since that is governed by when they are first needed (which simplifies understanding of startup/shutdown events: they just move the dependency creation to happen before the first request).

## Dependency Lifespans

### Request
This is the current default / only option.
Once the endpoint function runs, the dependency is destroyed in the background.

### App
The original proposal in this PR.
The dependency gets destroyed when the app is destroyed.

### Endpoint
The dependency gets destroyed when the endpoint finishes running (i.e. immediately before the request is done processing). Notably, this would allow the dependency to raise `HTTPExceptions` while being torn down. This is has already been proposed and implemented in #2697.

### None/ Background
The dependency gets destroyed in the background immediately after its value is retrieved.
This could be useful if for example you retrieve a scalar value or something and need to do some teardown, but the endpoint doesn't depend on the teardown itself, it only needs the value.
Instead of waiting until the endpoint finishes running to start the teardown (which is the behavior with the current / `Request` lifespan), the teardown starts immediately after the the value is retrieved.

## Caching

### True / Request
This is the current default behavior.
The dependency's returned value is cached for the lifetime of the request.

### False
This is current optional behavior
Every time this dependency is requested, it is re created.

### App
Proposed in this PR.
The _value_ of the dependency is cached for the lifetime/lifespan of the App (note that the dependency itself could be torn down immediately, eg. if using `lifespan="background"`).
This acts as a default for request cached dependencies: if an app-cached value exists, it is returned and the dependency is not re-constructed.

**Comment by adriangb** [comment]
I pushed a series of commits implementing the options proposed in https://github.com/tiangolo/fastapi/pull/3516#issuecomment-885107031.

The only one I didn't implement is `lifetime="background"` because the implementation is dissimilar to the others and requires a bit more technical thought; that can be a future PR / feature.

I'd like to acknowledge @zoliknemet for the excellent implementation in #2697, which I partially borrowed to implement `lifetime="endpoint"`.

Additionally:
- The implementation is now using enums, as proposed in https://github.com/tiangolo/fastapi/pull/3516#issuecomment-884397298. `cache={True,False}` gets converted to the Enum names if passed, making the change backwards compatible (although it might break static analysis tools unless we add a `Union[...]` type hint).
- Any startup/shutdown dependencies that depend on `Request`, `Response`, etc. are caught and a user-friendly error is given.

File: fastapi/dependencies/utils.py
Comment (adriangb) [inline fastapi/dependencies/utils.py:543]: I hope this is a sensible set of exclusions. I dunno what to do with the security dependencies.

**Comment by adriangb** [comment]
Some more doc edits. I've gone with editing the existing examples referenced in https://github.com/tiangolo/fastapi/pull/3516#issuecomment-885334657 instead of duplicating them.

**Comment by adriangb** [comment]
In case any of your are interested, I made a toy repo where I've been prototyping what it would look like to generalize the DI system to be less request/response coupled and to support the features proposed in this PR.

Here's an example of how it might integrate into FastAPI internally (the goal is near zero user facing changes): https://github.com/adriangb/anydep/blob/main/comparisons/test_anydep_request_params.py

I was also able to get some other nice features working:
- Anyio compatibility
- Pretty good speedups for highly nested / branched dependencies by parallelizing execution using anyio task groups
- Full typing support (i.e. `Depends` tells mypy that it returns the type that the callable you gave it is annotated with instead of `Any`)

**Comment by adriangb** [comment]
I'm going to close this in favor of discussion in #3641

**Pull request closed without merging**
